lib/pacemaker/pcmk_output.c

```diff
@@ -2141,6 +2141,13 @@ attribute_default(pcmk__output_t *out, va_list args)
 
     GString *s = g_string_sized_new(50);
 
+    char *value_esc = NULL;
+
+    if (pcmk__xml_needs_escape(value, true)) {
+        value_esc = pcmk__xml_escape(value, true);
```

Comment:
Fabio is concerned that some agent or tool may depend on the literal `"\n"` (backslash followed by newline, NOT an escape character) that's in the output as of Pacemaker 2.1.7.

Prior to this commit, we output a literal newline (with line break) for text output objects if the attribute value has a newline, so the value gets split over multiple lines. This commit causes a newline to be printed as `"&#x0A;"` by `attrd_updater` and `crm_attribute` for text output objects.

That is my preference, because it's actually a meaningful character and because printing `"\n"` further complicates the implementation. (`pcmk__xml_escape()` never prints `"\n"` for newline anymore.) We could print `"\n"`, however.

Reply:
How were we escaping this to begin with, since this is not used for XML output?

Reply:
Good question. I don't see a direct call to `crm_xml_escape()`.

I'm assuming something is sending a query, getting an XML response, parsing it for some reason, getting the attr value, and printing it. That would explain the behavior in both cases (once something becomes "\n", it never goes back to a newline; and "&#x0A;" is parsed back to a newline).

Trying to trace how that might be happening.

Reply:
`pcmk__ipc_prepare_iov()` on the IPC client side dumps the request XML to a string before sending it. That turns the newline into whatever we use to escape a newline, if anything. Then we parse that string on the IPC server side before setting the attribute.

So a newline set by a CLI tool used to be escaped to `"\n"` before set time. In `main`, it doesn't get escaped, so the dump/parse sequence transforms it to a single space. As of this commit, it gets *properly* escaped, so the dump/parse sequence preserves the newline, and it's ultimately dumped as `"&#x0A;"`

Reply:
The XML output looks ideal now. The text output is most correct before "Escape special chars in XML attribute text output" (i.e. the text output is identical to what the user added), though of course OpenStack can't handle that. It doesn't make sense to put XML entities in text output, so I'm leaning to Fabio's recommendation to keep showing backslash-n in that case. We can add a COMPAT comment to eventually switch to the correct text output (hopefully we can get OpenStack to fix it to handle both formats, then switch at 3.1).

Reply:
> The text output is most correct before "Escape special chars in XML attribute text output" (i.e. the text output is identical to what the user added)

I agree

> backslash-n in that case

Alright, I'll do it as cleanly as I can. It will probably involve an extra strdup() in the "attribute" message formatter, since I don't want to add and maintain another argument to `pcmk__xml_escape()`. Should be fine since it's rarely called.

> hopefully we can get OpenStack to fix it to handle both formats, then switch at 3.1

Yeah it's a trivial fix within the agents. I demonstrated a fix in the agent that works for both formats without having to check feature set or anything. It hasn't been implemented because the push has been to "fix" Pacemaker.

Fabio is concerned that there may be other, currently unknown agents that depend on the `"\n"`. I feel like we've gotta make the switch at some point though.

He is also concerned that the recent changes in escaping (particularly within the XML) may affect rolling upgrades. I told him I don't think they would; once an attribute is set, its value won't change just because Pacemaker was upgraded to a version that escapes differently. Let me know if you feel differently. The only possible edge case that comes to mind is implicit resources, and even that seems unlikely to change based on escaping changes.

Reply:
> I don't want to add and maintain another argument to `pcmk__xml_escape()`

Still true but I think that might be the easier approach